feat: add address list (BIP-44)

gantunesr opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This PR adds a a new view to see all addresses spread across all available networks.

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: Adds address list for multichain accounts.

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/MUL-549

## **Manual testing steps**

```gherkin
Feature: Multichain account address list

  Scenario: user shares address
    Given an user that has multichain account
    When the user presses the copy icon in the header
    Then navigates to a list of all available address in the account
```

**Multichain accounts feature flag should be enabled.**

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

https://github.com/user-attachments/assets/188dc5f2-ee21-497a-8fca-3da966c4b63c

### **After**

https://github.com/user-attachments/assets/aebf605c-b2fe-47cf-b1e2-65313b994c19

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

gantunesr committed:
chore: remove state 1 ff from AccountTreeInitService

gantunesr committed:
wip: address list UI

gantunesr committed:
chore: add styles

gantunesr committed:
Merge branch 'main' of https://github.com/MetaMask/metamask-mobile into gar/feat/address-list-ui

gantunesr committed:
chore: update type in component

gantunesr committed:
Merge branch 'main' of https://github.com/MetaMask/metamask-mobile into gar/feat/address-list-ui

gantunesr committed:
chore: add selectors

gantunesr committed:
feat: update AddressCopy component

gantunesr committed:
chore: add selectSelectedAccountGroupId selector

gantunesr committed:
chore: revert BaseAccountDetails

gantunesr committed:
chore: remove selectGroupIdByAccountId

gantunesr committed:
chore: improve AddressList

gantunesr committed:
Merge branch 'main' into gar/feat/address-list-ui

gantunesr commented on app/components/Views/MultichainAccounts/AddressList/AddressList.tsx:
```diff
@@ -0,0 +1,75 @@
+        <FlashList
+          data={internalAccountsSpreadByScopes}
+          keyExtractor={(item) => item.scope}
+          renderItem={renderAddressItem}
```
scope should be unique across the list since it is expected to have duplicated scopes in the data, if this is the case then there is a bug upstream

gantunesr committed:
Merge branch 'main' of https://github.com/MetaMask/metamask-mobile into gar/feat/address-list-ui

gantunesr committed:
Merge branch 'gar/feat/address-list-ui' of https://github.com/MetaMask/metamask-mobile into gar/feat/address-list-ui

gantunesr committed:
test: fix unit test

gantunesr committed:
test: add e2e IDs

owencraston commented on app/components/UI/AddressCopy/AddressCopy.tsx:
```diff
@@ -72,13 +82,28 @@ const AddressCopy = ({ account, iconColor, hitSlop }: AddressCopyProps) => {
+    });
+  };
+
+  const handleOnPress = () => {
```
should this be put into a useCallback?

gantunesr replied:
done in https://github.com/MetaMask/metamask-mobile/pull/18325/commits/86ff06e5bf043412416c933cf3edaef61c9af724

gantunesr committed:
Merge branch 'main' into gar/feat/address-list-ui

gantunesr committed:
test: fix mocks

gantunesr committed:
refactor: refactor AddressCopy

gantunesr committed:
fix: fix import

gantunesr committed:
fix: types

gantunesr committed:
fix: lint

gantunesr committed:
refactor: update AddressCopy to use useCallback hook

gantunesr committed:
Merge branch 'main' into gar/feat/address-list-ui

gantunesr committed:
refactor: refactor `MultichainAddressRow` (#18442)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Refactor the `MultichainAddressRow` to dynamically render icon+actions

## **Changelog**

CHANGELOG entry: null

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/MUL-404
(partially)

## **Manual testing steps**

N/A

## **Screenshots/Recordings**

N/A

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

wachunei commented on app/components/Views/MultichainAccounts/AddressList/types.ts:
```diff
@@ -0,0 +1,18 @@
+import { type CaipChainId } from '@metamask/utils';
+import { type InternalAccount } from '@metamask/keyring-internal-api';
+
+export interface Props {
+  route: {
+    params: {
+      groupId: AccountGroupId;
+      title: string;
+    };
+  };
+}
```
This should be refactored as part of the createNavigationDetails helper.

gantunesr replied:
Done in https://github.com/MetaMask/metamask-mobile/commit/89240af525384c6333b41eae6c44bd8eb7984411

wachunei commented on app/components/Views/MultichainAccounts/AddressList/AddressList.test.tsx:
since `AddressList` is meant to be a modal screen, `renderScreen` must be used instead for the tests.

gantunesr committed:
fix: update row icons

gantunesr committed:
Merge branch 'main' of https://github.com/MetaMask/metamask-mobile into gar/feat/address-list-ui

gantunesr committed:
Merge branch 'main' into gar/feat/address-list-ui

gantunesr committed:
refactor: refactor navigation options

gantunesr commented on app/component-library/components-temp/MultichainAccounts/MultichainAddressRow/MultichainAddressRow.tsx:
```diff
@@ -77,21 +65,16 @@ const MultichainAddressRow = ({
+            iconColor={IconColor.Default}
+            testID={icon.testId}
+          />
+        ))}
```
Will do in a separate PR

gantunesr commented on app/components/Nav/App/App.tsx:
```diff
@@ -710,6 +711,25 @@ const MultichainWalletDetails = () => {
+        animationEnabled: false,
+      }}
+    >
+      <Stack.Screen
+        name={Routes.MULTICHAIN_ACCOUNTS.ADDRESS_LIST}
+        component={AccountAddressList}
+        initialParams={route?.params}
+      />
```
Will refactor the nav for multichain account views in a separate PR

gantunesr commented on app/components/UI/AddressCopy/AddressCopy.tsx:
```diff
@@ -70,15 +73,40 @@ const AddressCopy = ({ account, iconColor, hitSlop }: AddressCopyProps) => {
+    navigate(Routes.MULTICHAIN_ACCOUNTS.ADDRESS_LIST, {
+      groupId: selectedAccountGroupId,
+      title: `${strings('multichain_accounts.address_list.receiving_address')}`,
+    });
```
Done in https://github.com/MetaMask/metamask-mobile/pull/18325/commits/89240af525384c6333b41eae6c44bd8eb7984411

gantunesr commented on app/components/Views/MultichainAccounts/AddressList/AddressList.tsx:
```diff
@@ -0,0 +1,76 @@
+  const navigation = useNavigation();
+  const { styles } = useStyles(styleSheet, {});
+
+  const { groupId, title } = props.route.params;
```
Done in https://github.com/MetaMask/metamask-mobile/commit/89240af525384c6333b41eae6c44bd8eb7984411

gantunesr commented on app/component-library/components-temp/MultichainAccounts/MultichainAddressRow/MultichainAddressRow.types.ts:
```diff
@@ -14,6 +33,13 @@ export interface MultichainAddressRowProps {
+   * The callback will be executed when the icon is pressed.
+   * Icons are displayed in the order they are provided.
+   */
+  icons: Icon[];
```
Will do in a separate PR

gantunesr committed:
Merge branch 'main' into gar/feat/address-list-ui

owencraston commented on app/components/Views/MultichainAccounts/AddressList/AddressList.tsx:
```diff
@@ -0,0 +1,107 @@
+    selectInternalAccountsSpreadByScopes(groupId);
+
+  const copyAddressToClipboard = useCallback((address: string) => {
+    ClipboardManager.setString(address);
```
should this be awaited?

owencraston replied:
Secondly, should we port over the confirmation logic that existed on the old AddressCopy screen so that the user has feedback that they copied the address? Currently it is not obvious to me that the copy worked.

<img width="634" height="494" alt="Screenshot 2025-08-21 at 4 06 25 PM" src="https://github.com/user-attachments/assets/df4c486a-7065-4a87-b389-b3894598fdaa" />

gantunesr replied:
await done in [31c55d0](https://github.com/MetaMask/metamask-mobile/pull/18325/commits/31c55d0ee1ad1c4dc6a89ffad525c52add439e57)

gantunesr committed:
fix: set setString as async

gantunesr committed:
Merge branch 'gar/feat/address-list-ui' of https://github.com/MetaMask/metamask-mobile into gar/feat/address-list-ui

gantunesr committed:
Merge branch 'main' into gar/feat/address-list-ui
